backend/src/server.rs

```diff
@@ -0,0 +1,506 @@
+
+    let kind = body["kind"].as_str().unwrap();
+    if kind != "rustc" && kind != "benchmarks" {
+        return Err(format!("bad value kind: {}", kind).into());
+    }
+
+    let date = get_date_from_body(&body)?;
+    let data = data.by_kind(kind);
+    let day = &data[end_idx(data, &date)];
+    let mut by_crate = JsonValue::new_object();
+    for (crate_name, krate) in &day.by_crate {
+        let mut krate_obj = JsonValue::new_object();
+        for (phase_name, timing) in krate {
+            krate_obj[phase_name] = object! {
+                "percent" => timing.percent,
+                "time" => timing.time,
+                "rss" => timing.rss
+            };
+        }
+
+        by_crate[crate_name] = krate_obj;
+    }
+
+    Ok(object! {
+        "date" => day.date.format(JS_DATE_FORMAT).to_string(),
+        "commit" => day.commit.clone(),
+        "data" => by_crate
+    })
+}
+
+// Expected fields on body: {
+//     kind: 'rustc' | 'benchmarks',
+//     dates: [Date],
+//     crates: [str],   // crate == benchmarks for benchmark mode
+//     phases: [str],
+//     group_by: 'crate' | 'phase'
+// }
+// crate or phase can be 'total'
+fn get_days(req_body: Option<JsonValue>, data: &InputData) -> Result<JsonValue> {
+    let body = assert_request_body_present(req_body)?;
+
+    let kind = body["kind"].as_str().unwrap();
+    if kind != "rustc" && kind != "benchmarks" {
+        return Err(format!("bad value kind: {}", kind).into());
+    }
+    let group_by = body["group_by"].as_str().unwrap();
+    if group_by != "crate" && group_by != "phase" {
+        return Err(format!("bad value group_by: {}", group_by).into());
+    }
+
+    let mut result = JsonValue::new_array();
+    for orig_date in body["dates"].members() {
+        let date = match handle_date(orig_date.as_str().unwrap()) {
+            Ok(date) => date,
+            Err(err) => {
+                println!("bad date {:?}: {:?}", orig_date, err);
+                continue;
+            }
+        };
+        let data = data.by_kind(kind);
+        let day = get_data_for_date(&data[end_idx(data, &date)], &body["crates"], &body["phases"], group_by);
+        result.push(day)?;
+    }
+    Ok(result)
+}
+
+// Expected fields on body: {
+//     kind: 'rustc' | 'benchmarks',
+//     start: Date,     // optional
+//     end: Date,       // optional
+//     crates: [str],   // crate == benchmarks for benchmark mode
+//     phases: [str]
+// }
+// crate (rustc only) or phase can be 'total'
+fn get_stats(req_body: Option<JsonValue>, data: &InputData) -> Result<JsonValue> {
+    let body = assert_request_body_present(req_body)?;
+
+    let kind = body["kind"].as_str().unwrap();
+    if kind != "rustc" && kind != "benchmarks" {
+        return Err(format!("bad value kind: {}", kind).into());
+    }
+
+    if kind == "benchmarks" && body["crates"].contains("total") {
+        return Err(format!("bad value for crates {:?} with kind benchmarks", body["crates"]).into());
```

Comment:
Any ideas as to why a "benchmarks" kind can't have "total" in the crates? I'm a little confused by this.